### DISCUSSION TITLE
Top-level await

### DISCUSSION THREAD

**Discussion by Norcim133** [whitphx/stlite#1164] [Q&A] [answered]
I have a large B2B app and am loving stlite for its deployment.

Does anyone know a way to handle the following constraints?
1. To run the app in vanilla streamlit, I can't have a top-level await (it errors even if I place it in an if-statement based on env)
2. To use a top-level await when deployed in stlite, I tried exec"""await func...""" but I still get an error that way
3. I can kind of cheese a solution in stlite with asyncio.get_event_loop and run_until_completion... but I know that could have unpredictable impact on event loops
4. When I do #3, I can't get streamlit to update on the return from the loop, even if its in a try/finally... I have to manually add a delay and trigger a re-run

In summary, does anyone know how to do a top-level await that would work during vanilla streamlit (like wrapping main in asyncio.run(main)... but which would also work when deployed in stlite?

**Comment by whitphx** [comment]
Hi,

> To use a top-level await when deployed in stlite, I tried exec"""await func...""" but I still get an error that way

I couldn't get this. `async def` + top-level `await` should work on Stlite like [this demo](https://edit.share.stlite.net/#!ChBzdHJlYW1saXRfYXBwLnB5EsUBChBzdHJlYW1saXRfYXBwLnB5ErABCq0BaW1wb3J0IHN0cmVhbWxpdCBhcyBzdAppbXBvcnQgYXN5bmNpbwoKYXN5bmMgZGVmIGZvbygpOgogICAgc3Qud3JpdGUoIlN0YXJ0IikKICAgIHdpdGggc3Quc3Bpbm5lcigpOgogICAgICAgIGF3YWl0IGFzeW5jaW8uc2xlZXAoMSkKICAgICAgICBzdC53cml0ZSgiQXdhaXRlZCIpCgphd2FpdCBmb28oKQo,). What does your code look like?

> does anyone know how to do a top-level await that would work during vanilla streamlit

alaik it's not possible on the normal Streamlit now.

You mean, you want the full code portability between Streamlit and Stlite when the code contains async & await?

**Comment by Norcim133** [comment]
Hey!

My code is complicated so here
<https://share.stlite.net/#!Cg5leGFtcGxlX2FwcC5weRLFAQoQc3RyZWFtbGl0X2FwcC5weRKwAQqtAWltcG9ydCBzdHJlYW1saXQgYXMgc3QKaW1wb3J0IGFzeW5jaW8KCmFzeW5jIGRlZiBmb28oKToKICAgIHN0LndyaXRlKCJTdGFydCIpCiAgICB3aXRoIHN0LnNwaW5uZXIoKToKICAgICAgICBhd2FpdCBhc3luY2lvLnNsZWVwKDEpCiAgICAgICAgc3Qud3JpdGUoIkF3YWl0ZWQiKQoKYXdhaXQgZm9vKCkKEpUFCg5leGFtcGxlX2FwcC5weRKCBQr_BGltcG9ydCBzdHJlYW1saXQgYXMgc3QKaW1wb3J0IGFzeW5jaW8KCiMgV2Ugd2FudCB0byBydW4gaW4gZWl0aGVyIGVudmlyb25tZW50CmVudl9vcHRpb25zID0gWyJzdHJlYW1saXQiLCAic3RsaXRlIl0KCiMgV2UgY2hvb3NlIHN0cmVhbWxpdApjaG9zZW5fZW52ID0gZW52X29wdGlvbnNbMV0KCiMgT3JpZ2luYWwgc3luYyBmdW5jCmFzeW5jIGRlZiBmb28oKToKICAgIHN0LndyaXRlKCJTdGFydCIpCiAgICB3aXRoIHN0LnNwaW5uZXIoKToKICAgICAgICBhd2FpdCBhc3luY2lvLnNsZWVwKDEpCiAgICAgICAgc3Qud3JpdGUoIkF3YWl0ZWQiKQoKCmlmIGNob3Nlbl9lbnYgPT0gInN0bGl0ZSI6CiAgICBhd2FpdCBmb28oKSAjU3RyZWFtbGl0OiAiU3ludGF4RXJyb3I6IGF3YWl0IGlzIG9ubHkgdmFsaWQgaW4gYXN5bmMgZnVuY3Rpb24iCgplbGlmIGNob3Nlbl9lbnYgPT0gInN0cmVhbWxpdCI6CiAgICBhc3luY2lvLnJ1bihmb28oKSkgI1N0bGl0ZTogIlJ1bnRpbWVFcnJvcjogYXN5bmNpby5ydW4oKSBjYW5ub3QgYmUgY2FsbGVkIGZyb20gYSBydW5uaW5nIGV2ZW50IGxvb3AiCgojIEV2ZW4gdGhvdWdoIHdlIHRyeSB0byBjb25kaXRpb25hbGx5IGlnbm9yZSBhd2FpdCAmIGFzeW5jaW8ucnVuLCB3ZSBnZXQgZXJyb3JzCg,=>
is
an example that illustrates the challenge.

I want code to work in both streamlit and stlite. For example:

> if stlite_env:
>     await foo()
> *<==In streamlit: SyntaxError*elif streamlit_env:
>     asyncio.run(foo()) *<==In stlite: RuntimeError*


Even though they are in conditional, the other environment still raises an
error.

The problem is, even in a try/except block, streamlit will give syntax error

> try:
>     await foo()
> *<==In streamlit: still gives SyntaxError*except:
>     asyncio.run(foo()) *<==In stlite: works*


And if I try to do it in *if __name__ == "__main__ *it is same issue just
one level higher.

**Reply by whitphx** [reply]
Thanks, I understood.
I think it's not possible for now.

Let's track this issue on #1173 where I will try to make `asyncio.run()` work on Stlite, which should help this.

**Comment by Norcim133** [comment]
Thanks!

In case it is ever helpful, I technically got it working this way:
a) put streamlit version in asyncio.run(my_func())

b) put stlite version in this:
loop = asyncio.get_event_loop()
try:
    loop.run_until_complete(my_func())
finally:
    loop.close()
    <re-render updated widgets>

The problem is the "finally" wouldn't actually wait until "try" completed.
So then the re-render would be missed and you would have to manually re-run
to get the outcomes of the function. Adding a delay made it seem to work.

**Comment by whitphx** [comment, marked as answer]
@Norcim133 
I'm working on it.
Do you think this helps?
https://31ad5dfc.stlite-sharing-editor.pages.dev/#!ChBzdHJlYW1saXRfYXBwLnB5EswBChBzdHJlYW1saXRfYXBwLnB5ErcBCrQBaW1wb3J0IHN0cmVhbWxpdCBhcyBzdAppbXBvcnQgYXN5bmNpbwoKYXN5bmMgZGVmIGZvbygpOgogICAgc3Qud3JpdGUoIlN0YXJ0IikKICAgIHdpdGggc3Quc3Bpbm5lcigpOgogICAgICAgIGF3YWl0IGFzeW5jaW8uc2xlZXAoMSkKICAgICAgICBzdC53cml0ZSgiQXdhaXRlZCIpCgphc3luY2lvLnJ1bihmb28oKSkK
This is a prototype env.

**Reply by Norcim133** [reply]
That should totally work. My app is too big to test fully there but that top asyncio.run is all I need.

It's a nice approach because it means you can test a big app in streamlit without having to wait for deployment and load over and over.

**Reply by whitphx** [reply]
Thanks. This new feature is out since v0.72.0

**Reply by Norcim133** [reply]
I tried it and can confirm it works. That makes life much easier. Amazing!
